File: hackerrank/problem solving/algorithms/2 - implementation/28 - jumping on the clouds - revisited/oo/jumping_on_the_clouds_revisited_oo.rb
# Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds-revisited/problem?isFullScreen=true

def main
    array = readAnArray
    sizeArray = array.first
    jumpLength = array.last
    array = readAnArray

    obj = JumpingOnTheClouds.new array, jumpLength
    puts obj.getRemainingEnergy
end

    def readAnArray
        array = gets.split.map(&:to_i)
    end

    class JumpingOnTheClouds
        @array
        @jumpLength
        @energy

        def initialize array, jumpLength
            @array = array
            @jumpLength = jumpLength
            @energy = 100

            _jumpOnClouds
        end

            private def _jumpOnClouds
                cloudIndex = 0
                while true
                    @energy -= _spentEnergyAccordingToTypeOfCloud @array[cloudIndex]
                    cloudIndex = _generateNewCloudIndex cloudIndex
                    if _isCloudIndexBackToFirstCloud cloudIndex
                        break
                    end
                end
            end

                private def _spentEnergyAccordingToTypeOfCloud cloudType
                    cloudType == 0 ? 1 : 3
                end

                private def _generateNewCloudIndex cloudIndex
                    (cloudIndex + @jumpLength) % @array.size
                end

                private def _isCloudIndexBackToFirstCloud cloudIndex
                    cloudIndex == 0
                end

        def getRemainingEnergy
            @energy
        end
    end

main

File: hackerrank/problem solving/algorithms/2 - implementation/28 - jumping on the clouds - revisited/oo/jumping_on_the_clouds_revisited_oo.cs
// Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds-revisited/problem?isFullScreen=true

using System;


public class Solution
{
    public static void Main()
    {
        List<int> array = _ReadAnArray();
        int _sizeArray = array.First();
        int jumpLength = array.Last();
        array = _ReadAnArray();

        JumpingOnTheClouds obj = new JumpingOnTheClouds(array, jumpLength);
        Console.WriteLine(obj.GetRemainingEnergy());
    }

        private static List<int> _ReadAnArray()
        {
            List<int> array = Console.ReadLine().Split().Select(int.Parse).ToList();
            return array;
        }
}

    class JumpingOnTheClouds
    {
        List<int> _Array;
        int _JumpLength;
        int _Energy;

        public JumpingOnTheClouds(List<int> array, int jumpLength)
        {
            _Array = array;
            _JumpLength = jumpLength;
            _Energy = 100;

            _JumpOnClouds();
        }

            private void _JumpOnClouds()
            {
                for (int cloudIndex = 0; true; )
                {
                    _Energy -= _SpentEnergyAccordingToTypeOfCloud(_Array[cloudIndex]);
                    cloudIndex = _GenerateNewCloudIndex(cloudIndex);
                    if (_IsCloudIndexBackToFirstCloud(cloudIndex))
                        break;
                }
            }

                private int _SpentEnergyAccordingToTypeOfCloud(int cloudType)
                {
                    return cloudType == 0 ? 1 : 3;
                }

                private int _GenerateNewCloudIndex(int cloudIndex)
                {
                    return (cloudIndex + _JumpLength) % _Array.Count;
                }

                private bool _IsCloudIndexBackToFirstCloud(int cloudIndex)
                {
                    return cloudIndex == 0;
                }

        public int GetRemainingEnergy()
        {
            return _Energy;
        }
    }
